### PR TITLE
Allow for Bean Introspection to use different accessor for read and write. 

### PR DESCRIPTION
**Pull request by dstepanov**
Refactoring the introspection visitor to use the same bean properties logic as the rest of Micronaut. Added support for it to use different types of accessors. Now getter and setter don't need to share the same accessor style.

`@BeanProperties` has to be with retention CLASS instead of SOURCE, because Javac would not return source annotations, added the same exclude for CLASS retention as for SOURCE, even so, in the `DefaultAnnotationMetadata` the field is still called source.

PR is targeting my other branch.

### CODE REVIEW

File: inject-groovy/src/test/groovy/io/micronaut/inject/annotation/SourceRetentionSpec.groovy [hunk 1]
```diff
@@ -1,28 +0,0 @@
-package io.micronaut.inject.annotation
-
-import io.micronaut.ast.transform.test.AbstractBeanDefinitionSpec
-import io.micronaut.inject.BeanDefinition
-
-import java.lang.annotation.Native
-
-class SourceRetentionSpec extends AbstractBeanDefinitionSpec {
-    void "test source retention annotations are not retained"() {
```
Comment (graemerocher) [inline inject-groovy/src/test/groovy/io/micronaut/inject/annotation/SourceRetentionSpec.groovy:9]: why was this deleted?

File: inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataWriter.java [hunk 2]
```diff
@@ -487,7 +487,14 @@ private static void instantiateInternal(
         // 4th argument: all annotations
         pushCreateAnnotationData(owningType, declaringClassWriter, generatorAdapter, annotationMetadata.allAnnotations, defaultsStorage, loadTypeMethods, annotationMetadata.getSourceRetentionAnnotations());
         // 5th argument: annotations by stereotype
-        pushStringMapOf(generatorAdapter, annotationMetadata.annotationsByStereotype, false, Collections.emptyList(), list -> pushListOfString(generatorAdapter, list));
+        Map<String, List<String>> annotationsByStereotype = annotationMetadata.annotationsByStereotype;
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataWriter.java:490]: should this create a copy before writing? Otherwise you are mutating internal state?

Comment (dstepanov) [reply inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataWriter.java:490]: I make a copy a few lines below

File: inject-groovy/src/test/groovy/io/micronaut/inject/annotation/SourceRetentionSpec.groovy (hunk 1 above)
Comment (dstepanov) [reply inject-groovy/src/test/groovy/io/micronaut/inject/annotation/SourceRetentionSpec.groovy:9]: It’s renamed to RetentionSpec git diff a bit glitchy

**Review by graemerocher — APPROVED** [review]
